Sources/SwiftParser/Declarations.swift

```diff
@@ -2233,3 +2189,59 @@ extension Parser {
     }
   }
 }
+
+extension Parser {
+  mutating func parseAllEffectsSpecifiers() -> [SpecifierWithToken] {
+    var specifiers = [SpecifierWithToken]()
+    var loopProgress = LoopProgressCondition()
+    while let (specifier, handle) = self.canRecoverTo(anyIn: EffectsSpecifier.self), loopProgress.evaluate(currentToken) {
+      let (unexpected, token) = self.eatKeepUnexpected(handle)
+      specifiers.append(contentsOf: unexpected.map { SpecifierWithToken(specifier: nil, token: $0) })
+      specifiers.append(SpecifierWithToken(specifier: specifier, token: token))
+    }
+    return specifiers
+  }
+
+  func addMissingEffectsSpecifiers(atDecl: Bool, from unexpectedSpecifiers: [SpecifierWithToken], asyncOrReasync: inout RawTokenSyntax?, throwsOrRethrows: inout RawTokenSyntax?) {
+    for unexpected in unexpectedSpecifiers {
+      guard asyncOrReasync == nil || throwsOrRethrows == nil else {
+        break
+      }
+
+      if asyncOrReasync == nil, unexpected.specifier == .asyncContextualKeyword || (atDecl && unexpected.specifier == .reasyncContextualKeyword) {
+        asyncOrReasync = RawTokenSyntax(missing: unexpected.token.tokenKind, text: unexpected.token.tokenText, arena: self.arena)
+      }
+
+      if throwsOrRethrows == nil, unexpected.specifier == .throwKeyword || (atDecl && unexpected.specifier == .rethrowsKeyword) {
+        throwsOrRethrows = RawTokenSyntax(missing: unexpected.token.tokenKind, arena: self.arena)
+      }
+    }
+  }
+
+  mutating func parseEffectsSpecifiers(atDecl: Bool) -> (asyncOrReasync: RawTokenSyntax?,
+                                                         throwsOrRethrows: RawTokenSyntax?,
+                                                         specifiersAfterThrows: [SpecifierWithToken]){
+    var initialSpecifiers = parseAllEffectsSpecifiers()
+
+    var asyncOrReasync: RawTokenSyntax?
+    var throwsOrRethrows: RawTokenSyntax?
+    if !initialSpecifiers.isEmpty {
+      let first = initialSpecifiers.removeFirst()
```

Comment:
```suggestion
    if !initialSpecifiers.isEmpty, let first = initialSpecifiers.removeFirst()
```

Reply:
`removeFirst` isn't optional, so you can't have the let binding in the if. I'm guessing you're thinking of `dropFirst`, but we're modifying the underlying collection here.